network/src/tests.rs

```diff
@@ -0,0 +1,456 @@
+
+	(
+		JustifiedBlockAnnounceValidator::new(authorities, client.clone()),
+		client,
+	)
+}
+
+fn default_header() -> Header {
+	Header {
+		number: Default::default(),
+		digest: Default::default(),
+		extrinsics_root: Default::default(),
+		parent_hash: Default::default(),
+		state_root: Default::default(),
+	}
+}
+
+fn make_gossip_message_and_header(
+	client: Arc<TestApi>,
+	relay_chain_leaf: H256,
+) -> (GossipMessage, Header) {
+	let key = Sr25519Keyring::Alice.pair().into();
+	let signing_context = client
+		.runtime_api()
+		.signing_context(&BlockId::Hash(relay_chain_leaf))
+		.unwrap();
+	let header = default_header();
+	let candidate_receipt = AbridgedCandidateReceipt {
+		head_data: header.encode().into(),
+		..AbridgedCandidateReceipt::default()
+	};
+	let statement = Statement::Candidate(candidate_receipt);
+	let signature = sign_table_statement(&statement, &key, &signing_context);
+	let sender = 0;
+	let gossip_statement = GossipStatement {
+		relay_chain_leaf,
+		signed_statement: SignedStatement {
+			statement,
+			signature,
+			sender,
+		},
+	};
+
+	(GossipMessage::Statement(gossip_statement), header)
+}
+
+#[test]
+fn valid_if_no_data() {
+	let mut validator = make_validator(Vec::new());
+	let header = default_header();
+
+	assert!(
+		matches!(validator.validate(&header, &[]), Ok(Validation::Success)),
+		"validating without data is always a success",
+	);
+}
+
+#[test]
+fn check_gossip_message_is_valid() {
+	let mut validator = make_validator(Vec::new());
+	let header = default_header();
+
+	let res = validator.validate(&header, &[0x42]).err();
+	assert!(
+		res.is_some(),
+		"only data that are gossip message are allowed"
+	);
+	assert!(matches!(
+		*res.unwrap().downcast::<ClientError>().unwrap(),
+		ClientError::BadJustification(x) if x.contains("must be a gossip message")
+	));
+}
+
+#[test]
+fn check_relay_parent_is_head() {
+	let (mut validator, client) = make_validator_and_client(Vec::new());
+	let relay_chain_leaf = H256::zero();
+	let (gossip_message, header) = make_gossip_message_and_header(client, relay_chain_leaf);
+	let data = gossip_message.encode();
+	let res = validator.validate(&header, data.as_slice());
+
+	assert_eq!(
+		res.unwrap(),
+		Validation::Failure,
+		"validation fails if the relay chain parent is not the relay chain head",
+	);
+}
+
+#[test]
+fn check_relay_parent_actually_exists() {
+	let (mut validator, client) = make_validator_and_client(Vec::new());
+	let relay_chain_leaf = H256::from_low_u64_be(42);
+	let (gossip_message, header) = make_gossip_message_and_header(client, relay_chain_leaf);
+	let data = gossip_message.encode();
+	let res = validator.validate(&header, data.as_slice()).err();
+
+	assert!(
+		res.is_some(),
+		"validation should fail if the relay chain leaf does not exist"
+	);
+	assert!(matches!(
+		*res.unwrap().downcast::<ClientError>().unwrap(),
+		ClientError::UnknownBlock(_)
+	));
+}
+
+#[test]
+fn check_relay_parent_does_not_fail() {
```

Comment:
It does not fail but the implementation checks for failing?

Reply:
yeah makes no sense, will rename that